Add documentation on Reference Entities (#1037)
akeneo/pim-docs: reference_entities/create_new_attribute_type.rst

Create a new Reference Entity Attribute type
============================================

.. note::

   Reference Entities feature is only available for the **Enterprise Edition**.

> This cookbook will present you how to create your own Attribute Type for Reference Entities.

Reviewer: ```suggestion
This cookbook will present how to create your own Attribute Type for Reference Entities.
```

Currently, there are 6 types of Attribute for Reference Entities:

- Text
- Image
- Record (*Link Record to another Record*)
- Record Collection (*Link Record to several Records*)

> - Option

Reviewer 2: Single Option

> - Option Collection

Reviewer 2: Multiple Option

Requirements
------------

During this chapter, we assume that you already created a new bundle to add your custom Reference Entity Attribute. Let's assume its namespace is ``Acme\\CustomBundle``.

> Create the Attribute

Reviewer 2: I would rewrite those two last section as follow, without giving too much details about the how.
```
To make the PIM support a custom attribute, there are two main steps:
1. Create the attribute model and the services that will let us create the attribute, and edit its properties.
2. Create the record's value related to this new attribute. 
```

--------------------

> - Domain Attribute

Author: TODO: We need to write the instructions

--------------------

For this tutorial, we will create a **custom simple metric attribute** type for reference entity.
So we'll be able to create this kind of attribute on any Reference Entity:

.. image:: ../_images/reference_entities/create_simple_metric_attribute.png
  :alt: Simple Metric as new attribute type

And to edit its common properties:

.. image:: ../_images/reference_entities/edit_simple_metric_attribute.png
  :alt: Edit Simple Metric attribute

.. note::

   **A small word on our architecture:**

   For the reference entities, we followed the Hexagonal architecture. So we split our classes in 3 different layers: Domain, Application & Infrastructure.

   - First we'll create **Domain** classes (the new custom ``SimpleMetricAttribute`` itself and its ``Value Object``)
   - Then **Application** classes (only Commands from CQRS principle (https://martinfowler.com/bliki/CQRS.html) and regular updaters)
   - And to finish **Infrastructure** classes (an Hydrator to hydrate our ``SimpleMetricAttribute`` coming from SQL)

   It's **not mandatory to respect this architecture** in your custom project.

1) Your new Custom Attribute Entity
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Let's start with our new custom Attribute. It must extends the ``\Akeneo\ReferenceEntity\Domain\Model\Attribute\AbstractAttribute``.

Reviewer: ```suggestion
Let's start with our new custom Attribute. It must extend the ``\Akeneo\ReferenceEntity\Domain\Model\Attribute\AbstractAttribute``.
```

    use Akeneo\ReferenceEntity\Domain\Model\ReferenceEntity\ReferenceEntityIdentifier;

    class SimpleMetricAttribute extends AbstractAttribute
    {
        public static function createSimpleMetric(
            AttributeIdentifier $identifier,
            ReferenceEntityIdentifier $referenceEntityIdentifier,
            AttributeCode $code,
            LabelCollection $labelCollection,
            AttributeOrder $order,
            AttributeIsRequired $isRequired,
            AttributeValuePerChannel $valuePerChannel,
            AttributeValuePerLocale $valuePerLocale
        ) {
            return new self(
                $identifier,
                $referenceEntityIdentifier,
                $code,
                $labelCollection,
                $order,
                $isRequired,
                $valuePerChannel,
                $valuePerLocale
            );
        }

        /**
         * {@inheritdoc}
         */
        protected function getType(): string
        {
            return 'simple_metric';
        }
    }

2) Create the attribute
^^^^^^^^^^^^^^^^^^^^^^^

Now that we have our Attribute class, we need to create classes to handle its creation.

> We'll need first to add the "Creation command", it needs to extend ``\Akeneo\ReferenceEntity\Application\Attribute\CreateAttribute\AbstractCreateAttributeCommand``.

Reviewer 3: We have already given a link to a cqrs explanation which is explaining well the commands IMHO : https://github.com/akeneo/pim-docs/pull/1037/files#diff-8366a4d40f1aa1ae597e92e405d029ffR45

        }
    }

And to register it:

.. code-block:: yaml

    # src/Acme/CustomBundle/Resources/config/services.yml

    services:
        acme.infrastructure.persistence.hydrator.attribute.simple_metric_attribute_hydrator:
            class: Acme\CustomBundle\Infrastructure\Persistence\Sql\Attribute\Hydrator\SimpleMetricAttributeHydrator
            arguments:
                - '@database_connection'
            tags:
                - { name: akeneo_referenceentity.attribute_hydrator }

Frontend Part of The New Attribute Type
---------------------------------------

To be able to create your brand new Simple Metric attribute on a Reference Entity, we need to add some code in the frontend part.

To do so, you can put all needed code in one single file but you can (and are encouraged) to split it into multiple
files if needed.

To keep this example simple, we will create everything in this file :

``src/Acme/CustomBundle/Resources/public/reference-entity/attribute/simple_metric.tsx``

.. note::

    If you create a new attribute type, Akeneo will need three things to manage it in the frontend:

>     - A **model**: a representation of your attribute, it's properties and overall behaviour

Reviewer: ```suggestion
    - A **model**: a representation of your attribute, its properties and overall behaviour
```

>     - A **model**: a representation of your attribute, it's properties and overall behaviour

Reviewer: ```suggestion
    - A **model**: a representation of your attribute, its properties and overall behaviour
```

>     - A **reducer**: to be able to know how to modify it's custom properties and react to the user intentions (see https://redux.js.org/)

Reviewer: ```suggestion
    - A **reducer**: to be able to know how to modify its custom properties and react to the user intentions (see https://redux.js.org/)
```

    - A **view**: as a React component to be able to render a user interface and dispatch events to the application

1) Model
^^^^^^^^

> The model of your custom attribute will contain the common properties of an attribute (code, labels, scope, etc) but also it's custom properties

Reviewer: ```suggestion
The model of your custom attribute will contain the common properties of an attribute (code, labels, scope, etc) but also its custom properties
```

and behaviours (*even if for now, our Simple Metric attribute doesn't have any*). To integrate it with the rest of the PIM, your attribute needs to implement the Attribute interface and provide a denormalizer.

This is the purpose of this section: provide a denormalizer capable of creating your custom attribute implementing Attribute interface.

.. code-block:: javascript

    /**
     * ## Import section
     *

>      * This is where sits your dependencies to external modules using the standard import method (see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import)

Reviewer: ```suggestion
     * This is where your dependencies to external modulesare, using the standard import method (see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import)
```

>      * The paths are absolute and the root is the web/bundles folder (at the root of your PIM project)

Reviewer: They aren't really absolute, more like relative to the web/bundles folder

          createCode(normalizedSimpleMetricAttribute.code),
          createLabelCollection(normalizedSimpleMetricAttribute.labels),
          normalizedSimpleMetricAttribute.value_per_locale,
          normalizedSimpleMetricAttribute.value_per_channel,
          normalizedSimpleMetricAttribute.order,
          normalizedSimpleMetricAttribute.is_required
        );
      }

      /**
       * The only method to implement here: the normalize method. Here you need to provide a serializable object (see https://developer.mozilla.org/en-US/docs/Glossary/Serialization)
       */
      public normalize(): NormalizedSimpleMetricAttribute {
        return {
          ...super.normalize(),
          type: 'simple_metric'
        };
      }
    }

    /**
     * The only required part of the file: exporting a denormalize method returning a custom attribute implementing Attribute interface
     */
    export const denormalize = ConcreteSimpleMetricAttribute.createFromNormalized;

2) Reducer
^^^^^^^^^^

Now that we have our attribute model in the frontend, we need to define our Reducer to know how to modify custom properties and react to the user intentions.

> We'll see later how to handle those custom properties, for now, it's gonna be a really simple reducer.

Reviewer: ```suggestion
We'll see later how to handle those custom properties, for now, it's going to be a really simple reducer.
```

.. code-block:: javascript

    /**

>      * Our custom attribute reducer needs to receive as input the normalized custom attribute, the code of the additional property and the value of the additional property.

Reviewer: ```suggestion
     * Our custom attribute reducer needs to receive the normalized custom attribute as input, the code of the additional property and the value of the additional property.
```

     * It returns the normalized custom attribute.
     */
    const simpleMetricAttributeReducer = (
      normalizedAttribute: NormalizedSimpleMetricAttribute,
      propertyCode: string
    ): NormalizedSimpleMetricAttribute => {
      switch (propertyCode) {
        default:
          break;
      }

      return normalizedAttribute;
    };

    /**
     * The only required part of the file: exporting the custom attribute reducer.
     */
    export const reducer = simpleMetricAttributeReducer;

3) View
^^^^^^^

> The last part we need to do, it's to create the React component to be able to render a user interface and dispatch events to the application (https://reactjs.org/docs/react-component.html).

Reviewer: ```suggestion
The last part we need to do is to create the React component to be able to render a user interface and dispatch events to the application (https://reactjs.org/docs/react-component.html).
```

Frontend Part of The New Attribute Type
---------------------------------------

To be able to create your brand new Boolean attribute on a Reference Entity, we need to add some code in the frontend part.

To do so, you can put all needed code in one single file but you can (and are encouraged) to split it into multiple
files if needed.

To keep this example simple, we will create everything in this file :

``src/Acme/CustomBundle/Resources/public/reference-entity/attribute/boolean.tsx``

If you create a new attribute type, Akeneo will need three things to manage it in the frontend:
 - A model: a representation of your attribute, it's properties and overall behaviour
 - A reducer: to be able to know how to modify it's custom properties and react to the user intentions (see https://redux.js.org/)
 - A view: as a React component to be able to render a user interface and dispatch events to the application

1) Model
^^^^^^^^

The model of your custom attribute will contains the common properties of an attribute (code, labels, scope, etc) but also it's custom properties

> and behaviours. To interface it with the rest of the PIM, your attribute needs to implement the Attribute interface and provide a denormalizer.

Reviewer: ```suggestion
and behaviours. To integrate it with the rest of the PIM, your attribute needs to implement the Attribute interface and provide a denormalizer.
```

> and behaviours. To interface it with the rest of the PIM, your attribute needs to implement the Attribute interface and provide a denormalizer.

Reviewer: Do we define 'denormalizer' somewhere ?

``src/Acme/CustomBundle/Resources/public/reference-entity/attribute/boolean.tsx``

If you create a new attribute type, Akeneo will need three things to manage it in the frontend:

>  - A model: a representation of your attribute, those properties and overall behaviour

Reviewer: 'it's' made more sense
